In [1]:
import numpy as np

In [2]:
# stop words that are not useful for classification
stop_words = ["a", "about", "above", "above", "across", "after", "afterwards", "again", "against", "all", "almost", "alone",
"along", "already", "also","although","always","am","among", "amongst", "amoungst", "amount",  "an", "and", "another", "any",
"anyhow","anyone","anything","anyway", "anywhere", "are", "around", "as",  "at", "back","be","became", "because","become",
"becomes", "becoming", "been", "before", "beforehand", "behind", "being", "below", "beside", "besides", "between", "beyond", 
"bill", "both", "bottom","but", "by", "call", "can", "cannot", "cant", "co", "con", "could", "couldnt", "cry", "de",
"describe", "detail", "do", "done", "down", "due", "during", "each", "eg", "eight", "either", "eleven","else", "elsewhere",
"empty", "enough", "etc", "even", "ever", "every", "everyone", "everything", "everywhere", "except", "few", "fifteen", "fify"
, "fill", "find", "fire", "first", "five", "for", "former", "formerly", "forty", "found", "four", "from", "front", "full", 
"further", "get", "give", "go", "had", "has", "hasnt", "have", "he", "hence", "her", "here", "hereafter", "hereby", "herein"
, "hereupon", "hers", "herself", "him", "himself", "his", "how", "however", "hundred", "ie", "if", "in", "inc", "indeed",
"interest", "into", "is", "it", "its", "itself", "keep", "last", "latter", "latterly", "least", "less", "ltd", "made",
"many", "may", "me", "meanwhile", "might", "mill", "mine", "more", "moreover", "most", "mostly", "move", "much", "must"
, "my", "myself", "name", "namely", "neither", "never", "nevertheless", "next", "nine", "no", "nobody", "none", "noone",
"nor", "not", "nothing", "now", "nowhere", "of", "off", "often", "on", "once", "one", "only", "onto", "or", "other", "others"
, "otherwise", "our", "ours", "ourselves", "out", "over", "own","part", "per", "perhaps", "please", "put", "rather", "re",
"same", "see", "seem", "seemed", "seeming", "seems", "serious", "several", "she", "should", "show", "side", "since"
, "sincere", "six", "sixty", "so", "some", "somehow", "someone", "something", "sometime", "sometimes", "somewhere"
, "still", "such", "system", "take", "ten", "than", "that", "the", "their", "them", "themselves", "then", "thence",
"there", "thereafter", "thereby", "therefore", "therein", "thereupon", "these", "they", "thickv", "thin", "third", "this",
"those", "though", "three", "through", "throughout", "thru", "thus", "to", "together", "too", "top", "toward", "towards",
"twelve", "twenty", "two", "un", "under", "until", "up", "upon", "us", "very", "via", "was", "we", "well", "were", "what",
"whatever", "when", "whence", "whenever", "where", "whereafter", "whereas", "whereby", "wherein", "whereupon", "wherever",
"whether", "which", "while", "whither", "who", "whoever", "whole", "whom", "whose", "why", "will", "with", "within", 
"without", "would", "yet", "you", "your", "yours", "yourself", "yourselves", "the"]

In [3]:
import os
X = []
Y = []
for catg in os.listdir(r"C:\Users\sandhya\Desktop\Newsgroup"):           
    for doc in os.listdir(r"C:\Users\sandhya\Desktop\Newsgroup/"+catg):    
        with open(r"C:\Users\sandhya\Desktop\Newsgroup/"+catg+'/'+doc) as file: 
            X.append((doc,file.read()))
            Y.append(catg)
# X is the tuple having first element as file name and second as all words in the particular file 
# Y is the list of all classes as output

In [4]:
# splitting X and Y for training and testing
from sklearn import model_selection
X_train,X_test,Y_train,Y_test = model_selection.train_test_split(X,Y)

In [5]:
# buliding vocabulary all proper or meaningful words with key as words in text and value as their frequency
import re,string
vocab = {}
for i in range(len(X_train)):
    text = X_train[i][1]                  # from tuple X_train taking out text
    text = text.lower()                 # converting text to small letters
    text=re.split(r'\W+',text)           # removing punctuations
    for word in text:
    # dont add stop words or numbers or words less than 2
        if not(word.isalpha()) or word in stop_words or len(word)<=2:
            continue
    # if word not in vocab add the word to vocab and initialize the frequency to 1
        if word not in vocab:
                vocab[word] = 1
    # else increase the frequency of word by 1
        else:
                vocab[word] += 1
#vocab            

In [6]:
#Sorting the dictionary on basis of frequency of words in descending order
import operator
sorted_vocab = sorted(vocab.items(), key=operator.itemgetter(1),reverse=True)
feature1=sorted_vocab

In [7]:
#Plotting graph on no. of words vs frequency 
#On basis of graph we can decide the number of features we want to take
import matplotlib.pyplot as plt
final_feature=[]
feature2=[]
for i in range(len(feature1)):
    final_feature.append(i)
    feature2.append(feature1[i][1])
plt.plot(final_feature,feature2)
plt.axis([0,8000,1,5000])
plt.grid()
plt.show()   


<Figure size 640x480 with 1 Axes>

In [8]:
# Taking the top 2000 highest frequency words
final_feature = [feature1[i][0] for i in range(2000)]
#final_feature

In [9]:
# making the x_train dataset from the vocab in its usual form
X_train_final = np.zeros([len(X_train),len(final_feature)],int)
for i in range(len(X_train)):                 # itearting over X_train
    text = X_train[i][1]                      # from tuple X_train taking out text
    text = text.lower()                       # converting text to small letters so that there is ease in matching
    text = text.strip()                       # removing white spaces from strt & end
    text=re.split(r'\W+',text)              # removing punctuatios
    for word in text:                       # iterating over each word in text
   # if word is present in feature list then add 1 to the final X_train with index equal to index of the word in feature list
        if word in final_feature:  
                X_train_final[i,final_feature.index(word)] += 1

In [10]:
# making the x_test dataset from the vocab in its usual form
X_test_final = np.zeros((len(X_test),len(final_feature)))
for i in range(len(X_test)):             # itearting over X_train
    text = X_test[i][1]                   # from tuple X_train taking out text
    text = text.lower()                   # converting text to small letters so that there is ease in matching
    text = text.strip()                   # removing white spaces from strt & end
    text=re.split(r'\W+',text)          # removing punctuatios
    for word in text:                   # iterating over each word in text
#if word is present in feature list then add 1 to the final X_test with index equal to index of the word in feature list     
        if word in final_feature:
             X_test_final[i][final_feature.index(word)] += 1

In [11]:
print(X_train_final.shape)
print(len(Y_train))

(14997, 2000)
14997


In [12]:
# using sklearn multinomial naive bayes for predicting output classes
from sklearn.naive_bayes import MultinomialNB
clf = MultinomialNB() 
clf.fit(X_train_final,Y_train)
Y_pred = clf.predict(X_test_final)

In [13]:
from sklearn.metrics import classification_report ,confusion_matrix
print(classification_report(Y_test,Y_pred))
print(confusion_matrix(Y_test,Y_pred))

                          precision    recall  f1-score   support

             alt.atheism       0.79      0.81      0.80       250
           comp.graphics       0.92      0.75      0.82       248
 comp.os.ms-windows.misc       0.82      0.81      0.81       248
comp.sys.ibm.pc.hardware       0.71      0.77      0.74       232
   comp.sys.mac.hardware       0.78      0.88      0.83       231
          comp.windows.x       0.89      0.79      0.84       238
            misc.forsale       0.87      0.93      0.90       266
               rec.autos       0.88      0.95      0.91       242
         rec.motorcycles       0.88      0.95      0.92       264
      rec.sport.baseball       0.96      0.93      0.94       262
        rec.sport.hockey       0.95      0.94      0.95       246
               sci.crypt       0.97      0.95      0.96       229
         sci.electronics       0.78      0.90      0.84       256
                 sci.med       0.96      0.88      0.92       281
         

In [14]:
from sklearn.metrics import accuracy_score
print("Score on training data using sklearn :" , clf.score(X_train_final,Y_train))
print("Score on testing data using sklearn :" , clf.score(X_test_final,Y_test))

Score on training data using sklearn : 0.8941788357671534
Score on testing data using sklearn : 0.8622


In [15]:
def probability(dictionary,X,current_class):
    output = 0
 # getting the number of words having class as current class
    num_features = len(dictionary[current_class].keys())
        
    for j in range(num_features):   # iterating over each class in  output
    
    # count_of_current_class_with_value_xj is the number of words with class as current class
        count_of_current_class_with_value_xj = dictionary[current_class][final_feature[j]]   + 1
    # count_of_current_class is the total number of words in current class
        count_of_current_class = dictionary[current_class]["total"] +len(dictionary[current_class].keys())-1                       
     # probability_of_xj_for_given_class is the probability the a word belongs to current class
        probability_of_xj_for_given_class = np.log(count_of_current_class_with_value_xj) - (np.log(count_of_current_class))
    # total probability till the current class
        output = output + probability_of_xj_for_given_class
    return output    

In [16]:
def predictSinglePoint(dictionary,x):
    classes = dictionary.keys()       # classes has all the possible output classes 
    p_best_class = -1                 # assumption
    best_class = -1000                # assumption
    first_run = True                  # assumption
    for current_class in classes:       # iterating over each class
        if(current_class == "total"):           
            continue
        p_current_class = probability(dictionary,x,current_class)
        if(first_run or p_current_class > best_p):           
            best_p = p_current_class
            best_class = current_class
        first_run = False  
    return best_class    

In [17]:
def predict(dictionary,X_test):
    y_pred = []
    for x in X_test:                # iterating over X_test 
        x_class =  predictSinglePoint(dictionary,x)       # here x is a row in X_test i.e X_test[ith row]
        y_pred.append(x_class)                            # y_pred denotes the output list
    return y_pred

In [18]:
#final_feature

['edu',
 'cmu',
 'com',
 'news',
 'srv',
 'cantaloupe',
 'net',
 'message',
 'subject',
 'lines',
 'date',
 'apr',
 'newsgroups',
 'path',
 'organization',
 'state',
 'gmt',
 'ohio',
 'writes',
 'article',
 'comp',
 'references',
 'talk',
 'misc',
 'sender',
 'ans',
 'howland',
 'reston',
 'don',
 'like',
 'people',
 'university',
 'just',
 'posting',
 'nntp',
 'mps',
 'host',
 'know',
 'sci',
 'usenet',
 'zaphod',
 'think',
 'windows',
 'politics',
 'alt',
 'does',
 'time',
 'use',
 'rutgers',
 'harvard',
 'near',
 'eng',
 'new',
 'rec',
 'sei',
 'crabapple',
 'good',
 'noc',
 'soc',
 'xref',
 'das',
 'cis',
 'max',
 'world',
 'europa',
 'gtefsd',
 'ece',
 'religion',
 'way',
 'uunet',
 'god',
 'org',
 'christian',
 'make',
 'gov',
 'say',
 'right',
 'rochester',
 'distribution',
 'space',
 'club',
 'did',
 'mail',
 'andrew',
 'ibm',
 'want',
 'gatech',
 'sun',
 'magnesium',
 'said',
 'utexas',
 'reply',
 'used',
 'udel',
 'sys',
 'culture',
 'need',
 'work',
 'nasa',
 'uiuc',
 'reall

In [19]:
def fit(X,Y):
    Y = np.array(Y)
    classes = set(Y)                     # getting all the possible classes in output
    count = {} 
    for current_class in classes:          # iterating over each class
        temp = 0
        X_required = X[Y == current_class]       # all rows in X where class is current class

        count[current_class] = {}
        for word in final_feature:               # iterating over all words in feature list
           # count[current_class][word] has all words having class as current class
            count[current_class][word] = (X_required[:,final_feature.index(word)]).sum()
    
         # temp has all words having class as current class
        temp+= count[current_class][word]
        count[current_class]["total"] =  temp
        
   # print(final_feature)    
    return count    

In [20]:
def score(Y_test,Y_pred):
    count = 0
    for i in range(len(Y_pred)):
        if Y_test[i] == Y_pred[i]:
            count+=1
    ans = count/len(Y_pred)
    return ans

In [21]:
dictionary = fit(X_train_final,Y_train)
dictionary

{'comp.graphics': {'edu': 5551,
  'cmu': 1892,
  'com': 1310,
  'news': 1296,
  'srv': 1214,
  'cantaloupe': 999,
  'net': 1026,
  'message': 842,
  'subject': 820,
  'lines': 843,
  'date': 781,
  'apr': 673,
  'newsgroups': 777,
  'path': 780,
  'organization': 755,
  'state': 575,
  'gmt': 671,
  'ohio': 520,
  'writes': 284,
  'article': 309,
  'comp': 1782,
  'references': 371,
  'talk': 10,
  'misc': 93,
  'sender': 401,
  'ans': 436,
  'howland': 427,
  'reston': 431,
  'don': 306,
  'like': 345,
  'people': 122,
  'university': 463,
  'just': 242,
  'posting': 367,
  'nntp': 344,
  'mps': 320,
  'host': 362,
  'know': 304,
  'sci': 113,
  'usenet': 306,
  'zaphod': 297,
  'think': 150,
  'windows': 263,
  'politics': 0,
  'alt': 411,
  'does': 323,
  'time': 196,
  'use': 412,
  'rutgers': 14,
  'harvard': 259,
  'near': 274,
  'eng': 223,
  'new': 178,
  'rec': 38,
  'sei': 212,
  'crabapple': 213,
  'good': 190,
  'noc': 259,
  'soc': 0,
  'xref': 248,
  'das': 244,
  'cis': 

In [22]:
y_pred = predict(dictionary,X_test_final)

In [23]:
print(classification_report(Y_test,Y_pred))
print(confusion_matrix(Y_test,Y_pred))

                          precision    recall  f1-score   support

             alt.atheism       0.79      0.81      0.80       250
           comp.graphics       0.92      0.75      0.82       248
 comp.os.ms-windows.misc       0.82      0.81      0.81       248
comp.sys.ibm.pc.hardware       0.71      0.77      0.74       232
   comp.sys.mac.hardware       0.78      0.88      0.83       231
          comp.windows.x       0.89      0.79      0.84       238
            misc.forsale       0.87      0.93      0.90       266
               rec.autos       0.88      0.95      0.91       242
         rec.motorcycles       0.88      0.95      0.92       264
      rec.sport.baseball       0.96      0.93      0.94       262
        rec.sport.hockey       0.95      0.94      0.95       246
               sci.crypt       0.97      0.95      0.96       229
         sci.electronics       0.78      0.90      0.84       256
                 sci.med       0.96      0.88      0.92       281
         

In [24]:
print("Score on testing data own implementation :" , score(Y_test,Y_pred))

Score on testing data own implementation : 0.8622


 Result :
 with sklearn implementation score on testing data is 0.8622


 with own implementation score on testing data is 0.8622